Dev/paginated request class

zmaglica opened the pull request:
Fixes #5058 

#### Add Paginated request classes

New paginated request class have been created. Using this class, you can easily execute paginated API call to wc pay server.

Here is the example of paginated request:

```
 $page      = 2;
$page_size = 50;
$direction = 'asc';
$sort      = 'created';
$filters   = [ 'key'  => 'value' ];
$uri       = 'uri';

$request = Paginated::create();
$request->set_page( $page );
$request->set_page_size( $page_size );
$request->set_sort_direction( $direction );
$request->set_sort_by( $sort );
$request->set_filters( $filters );
$request->set_uri( $uri );
```
This class will return the same response from wc pay server. For simple requests where you need results from the server without formatting or modeling response data, use this paginated request. Since there could be plenty of similar requests where URI or URL is the only difference, the `set_uri` method provide an easy way to define API URI.

There are `List_Disputes` and `List_Transactions` request classes that are similar like paginated request class, but the only difference is that these classes format responses.

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Make sure that all tests are passing
* Test new request classes on your environment and make sure that everything is working as expected.

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

zmaglica committed:
Added create intent request

zmaglica committed:
Added create intent request class to UPE gateway

zmaglica committed:
Added update intent request class

zmaglica committed:
Integrated Update Intent class request

zmaglica committed:
Implemented get intent request

zmaglica committed:
Added get charge request

zmaglica committed:
Added charge request

zmaglica committed:
Code review fixes. Refactored create intent to create intentation

zmaglica committed:
Added constructors for Update Intent

zmaglica committed:
Implemented constructor intent id injection to Get Intent request

zmaglica committed:
Fixed tests and implemented update capture intent request

zmaglica committed:
Changed constructor params

zmaglica committed:
Fixed charged request

zmaglica committed:
Added paginated request

zmaglica committed:
Added base paginated requests

zmaglica committed:
Added list transactions class

zmaglica committed:
Added list disputed class

RadoslavGeorgiev commented on includes/core/server/class-request.php:
```diff
@@ -226,6 +228,26 @@ final public function send( $hook, ...$args ) {
+		try {
+			return $this->send( $hook, $args );
+		} catch ( API_Exception $e ) {
+			new WP_Error( $e->get_error_code(), $e->getMessage() );
```
`WP_Error`s are just objects, which are not thrown like exceptions, they need to be returned. And when returned, they still need to be handled. I don't understand how this method would help, let's see if the rest of the PR is going to clear that up.

zmaglica replied:
The purpose of this code is to mimic the behavior of this function: https://github.com/Automattic/woocommerce-payments/blob/develop/includes/admin/class-wc-payments-rest-controller.php#L48

Instead of getting an array and parse exception for each request, this function will do that for us inside the request class and only in one place.

RadoslavGeorgiev replied:
I should have revisited this comment :D

I see the point, let's rename it to something like `handle_rest_request` to make the purpose clear though, what do you think?

zmaglica replied:
Make sense. I will rename it to `handle_rest_request`

zmaglica replied:
The function name is renamed to `handle_rest_request`

RadoslavGeorgiev replied:
Alright. In any case, `WP_Error` is an object, which is not getting thrown in this case, and needs to be returned in order for the REST controller to handle it.

RadoslavGeorgiev commented on includes/admin/class-wc-rest-payments-charges-controller.php:
```diff
@@ -46,9 +46,8 @@ public function get_charge( $request ) {
-			$wcpay_request->set_charge_id( $charge_id );
-			$charge = $wcpay_request->send( 'wcpay_get_charge_request', $charge_id );
+			$wcpay_request = Get_Charge::create( $charge_id );
+			$charge        = $wcpay_request->send( 'wcpay_get_charge_request', $charge_id );
```
The charge ID is something that's already a parameter within the request, adding it to the filter here doesn't provide any benefit. Let's remove it.

zmaglica replied:
Removed.

RadoslavGeorgiev commented on includes/admin/class-wc-rest-payments-disputes-controller.php:
```diff
@@ -85,13 +87,14 @@ public function register_routes() {
+		$wcpay_request->set_filters( $this->get_disputes_filters( $request ) );
 
-		return $this->forward_request( 'list_disputes', [ $page, $page_size, $sort, $direction, $filters ] );
+		return $wcpay_request->handle_request( 'wcpay_list_disputes_request' );
```
Providing the `WP_Rest_Request $request` here could be really useful to this filter.

zmaglica replied:
Added.

RadoslavGeorgiev commented on includes/admin/class-wc-rest-payments-reader-controller.php:
```diff
@@ -279,9 +279,8 @@ public function generate_print_receipt( $request ) {
-			$charge_request->set_charge_id( $charge_id );
-			$charge_array = $charge_request->send( 'wcpay_get_charge_request', $charge_id );
+			$charge_request = Get_Charge::create( $charge_id );
+			$charge_array   = $charge_request->send( 'wcpay_get_charge_request', $charge_id );
```
Same point about the redundant `$charge_id` here.

zmaglica replied:
Removed.

RadoslavGeorgiev commented on includes/core/server/request/class-get-charge.php:
```diff
@@ -18,28 +19,17 @@
 
 	/**
-	 * Set intent id.
+	 * Sets the intent ID, which will be used in the request URL.
```
`intent` needs to be replaced here? Same with the parameter description.

zmaglica replied:
I will address this in intent branch.

RadoslavGeorgiev commented on includes/core/server/request/class-paginated.php:
```diff
@@ -0,0 +1,154 @@
+	 * @throws Invalid_Request_Parameter_Exception
+	 */
+	public function get_api(): string {
+		return $this->uri; // Will be changed on every request by overriding this method or use set_uri method.
```
I'm not sure I prefer the approach of a universal `Paginated` request here. Any reason against creating specific documents/deposits/authorizations classes? If everything they do is as simple as overwrite the `get_api()` method, it should be fairly easy to implement.

zmaglica replied:
I wanted to avoid creating class, import it, create test for it and use it only in one place. It felt to me like we are creating a bunch of classes where only one line of code will be changed. But, I see your point here. Adding extra class is fine approach too, but considering that we have around 100+ API calls, we should be careful in creating a bunch of classes since it will be difficult to navigate through the code.

RadoslavGeorgiev replied:
> since it will be difficult to navigate through the code

IDEs for the win :)

If it's just the API method that's changing, we could have a single test for each of those sub-requests within a single test class, and just add one line there to prevent coverage decreases.

zmaglica replied:
This is refactored and all classes or requests have their own class with dedicated URI and other things like filters and setters.

RadoslavGeorgiev commented on includes/core/server/request/class-paginated.php:
```diff
@@ -0,0 +1,154 @@
+	 * @return mixed|\WCPay\Core\Server\Response
+	 */
+	public function format_response( $response ) {
+		return $response;
```
Is there any reason to avoid the `Response` class here? I know it's this vs `to_array()`, and we'll be overwriting those responses with proper classes at some point in the future, so either option would be fine, but regressing like this in a sub-class feels weird.

zmaglica replied:
I wanted to avoid calling ->to_array() for every class. I will refactor this and place it inside `handle_request` method

zmaglica replied:
Moved into `hanndle_request` function

RadoslavGeorgiev commented on includes/core/server/request/class-paginated.php:
```diff
@@ -0,0 +1,154 @@
+	 *
+	 * @return void
+	 */
+	public function set_filters( array $filters ) {
```
Why not make the setters explicit, and add setters for individual filters within request classes?

zmaglica replied:
These filters have been added.

RadoslavGeorgiev commented on includes/admin/class-wc-rest-payments-authorizations-controller.php:
```diff
@@ -58,11 +60,14 @@ public function register_routes() {
-		$sort      = $request->get_param( 'sort' );
-		$direction = $request->get_param( 'direction' );
-		return $this->forward_request( 'list_authorizations', [ $page, $page_size, $sort, $direction ] );
+		$wcpay_request = Paginated_Request::create();
```
Considering that `uri`, `page`, `page_size`, `sort_by`, and `sort_direction` are universal, do you think we could create a non-final `create_from_rest_request` method within the `Paginated` class, something like this:

```php
class Paginated {
	public static function from_rest_request( $request ) {
		$wcpay_request = static::create();
		$wcpay_request->set_page( (int) $request->get_param( 'page' ) );
		$wcpay_request->set_page_size( (int) $request->get_param( 'pagesize' ) );
		$wcpay_request->set_sort_by( $request->get_param( 'sort' ) );
		$wcpay_request->set_sort_direction( $request->get_param( 'direction' ) );	
		return $wcpay_request;
	}
}
```
This way sub-classes will be able to nicely use filters as well:

```php
class List_Disputes extends Paginated {
	public static function from_rest_request( $request ) {
		$filters = [
			'match'           => $request->get_param( 'match' ),
			'currency_is'     => $request->get_param( 'store_currency_is' ),
			'created_before'  => $request->get_param( 'date_before' ),
			'created_after'   => $request->get_param( 'date_after' ),
			'created_between' => $request->get_param( 'date_between' ),
			'search'          => $request->get_param( 'search' ),
			'status_is'       => $request->get_param( 'status_is' ),
			'status_is_not'   => $request->get_param( 'status_is_not' ),
		];

		$wcpay_request = parent::from_rest_request( $request );
		$wcpay_request->set_filters( $filters );
		return $wcpay_request;
	}
}
```

By doing so we will simultaneously add a good base for a bridge between REST and server requests.

zmaglica replied:
All paginated request classes have this feature now.

RadoslavGeorgiev commented on includes/core/server/request/class-list-transactions.php:
```diff
@@ -0,0 +1,80 @@
+		'page'       => 0,
+		'pagesize'   => 25,
+		'sort'       => 'date',
+		'direction'  => 'desc',
```
Defaults are inherited from parents, at least `page` and `pagesize` do not to be re-defined here.

zmaglica replied:
This is fixed.

zmaglica committed:
Code review fixes. Fixed small nitpicks

zmaglica committed:
Used Response class to return data from class instrad of array and renamed handle request class to handle rest request class

zmaglica committed:
Added list depositsw request

zmaglica committed:
Added list documents request

zmaglica committed:
Added list authorizations request

zmaglica committed:
Fixed typos in test classes

zmaglica committed:
Fixed typos in request list classes

zmaglica committed:
Code review fixes. Fixed typos

RadoslavGeorgiev commented on includes/core/server/request/class-paginated.php:
```diff
@@ -38,7 +33,7 @@ class Paginated extends Request {
 	 */
 	public function get_api(): string {
-		return $this->uri; // Will be changed on every request by overriding this method or use set_uri method.
+		return ''; // Will be changed on every request that extends this class.
```
Does it make sense to keep this method? Leaving it abstract as the base request will enforce other paginated requests to implement it.

zmaglica replied:
Abstract function override is only allowed since PHP version 7.2. Currently, the minimum supported version of WooCommerce payment plugin is 7.0, so I didn't want to introduce a possible break change. We should bump minimum PHP supported version to 7.2 or leave it like this.

RadoslavGeorgiev replied:
Can't we just remove it from the `Paginated` class though? I don't think it needs to be overwritten.

zmaglica replied:
![image](https://user-images.githubusercontent.com/15913573/208458093-e41ef2b8-b9fe-4f8a-b921-f204bb3155b8.png)
PHPStorm cries about missing method in class, so I added it 😄 and I think that all child classes need to fill "body" of all abstract methods that are defined in parent abstract class (maybe I am wrong on this one, so I am not 100% certain)

RadoslavGeorgiev replied:
The class was used directly, but now that it's not, we should just mark it as abstract as well. Then there should be no more complaints.

zmaglica replied:
Good point. I really like approach. I make it abstract now.

zmaglica committed:
Added tests and custom filter setters for list disputes endpoint

zmaglica committed:
Added filter setters for list deposits request

zmaglica committed:
Added filter setters for list documents request

zmaglica committed:
Added filters for transaction class request and fixed date between param on all paginated requests

zmaglica committed:
Added capture intent request class

zmaglica committed:
Fixed typos in capture intent class

zmaglica committed:
Added cancel intent request class

zmaglica committed:
Added create setup intent request class

zmaglica committed:
Added create and confirm setup intent

RadoslavGeorgiev commented on includes/core/server/class-request.php:
```diff
@@ -226,6 +229,33 @@ final public function send( $hook, ...$args ) {
+	 */
+	final public function handle_rest_request( $hook, ...$args ) {
+		try {
+			$data = $this->send( $hook, $args );
```
Don't we need to expand `$args` here as well?
```suggestion
			$data = $this->send( $hook, ...$args );
```

zmaglica replied:
Fixed.

RadoslavGeorgiev commented on includes/core/server/request/class-paginated.php:
```diff
@@ -0,0 +1,140 @@
+		$wcpay_request->set_page( (int) $request->get_param( 'page' ) );
+		$wcpay_request->set_page_size( (int) $request->get_param( 'pagesize' ) );
+		$wcpay_request->set_sort_by( $request->get_param( 'sort' ) );
+		$wcpay_request->set_sort_direction( $request->get_param( 'direction' ) );
```
We need to check whether `page`, `pagesize`, `sort`, and `direction` are present within the REST request before setting them here. Removing one of them from the REST URL results in this right now:

```
Fatal error: Uncaught TypeError: Argument 1 passed to WCPay\Core\Server\Request\Paginated: :set_sort_by() must be of the type string,
null given, called in /var/www/html/wp-content/plugins/woocommerce-payments/includes/core/server/request/class-paginated.php on line 57 and defined in /var/www/html/wp-content/plugins/woocommerce-payments/includes/core/server/request/class-paginated.php on line 126
```

zmaglica replied:
This is fixed. Direction and sort is checked before calling set functions and page and page size have been cased to int with default values.

RadoslavGeorgiev commented on includes/core/server/request/class-paginated.php:
```diff
@@ -0,0 +1,140 @@
+
+
+	const DEFAULTS = [
+		'page'      => 0,
```
Shouldn't this default to 1, in case the parameter is omitted?

zmaglica replied:
If you take a look at `list_` endpoints in `WC_Payments_API_Client` class, you will find that the default value for page is zero. It felt wired to me too, but I kept it to zero because I didn't want to change previous behavior.

RadoslavGeorgiev replied:
It happens to also work with 0, so let's leave it that way then :)

RadoslavGeorgiev commented on includes/admin/class-wc-rest-payments-transactions-controller.php:
```diff
@@ -76,13 +78,16 @@ public function register_routes() {
+		$wcpay_request->set_sort_by( $request->get_param( 'sort' ) );
+		$wcpay_request->set_sort_direction( $request->get_param( 'direction' ) );
+		$wcpay_request->set_deposit_id( $request->get_param( 'deposit_id' ) );
+		$wcpay_request->set_filters( $this->get_transactions_filters( $request ) );
```
All of this is handled within the `List_Transactions` request already, let's remove it from this file.

zmaglica replied:
Removed.

RadoslavGeorgiev commented on includes/admin/class-wc-rest-payments-transactions-controller.php:
```diff
@@ -76,13 +78,16 @@ public function register_routes() {
+		$wcpay_request->set_page_size( (int) $request->get_param( 'pagesize' ) );
+		$wcpay_request->set_sort_by( $request->get_param( 'sort' ) );
+		$wcpay_request->set_sort_direction( $request->get_param( 'direction' ) );
+		$wcpay_request->set_deposit_id( $request->get_param( 'deposit_id' ) );
```
The deposit ID is not a part of `List_Transactions::from_rest_request` though, let's move it there before deleting the other lines here.

RadoslavGeorgiev commented on includes/core/server/request/class-list-transactions.php:
```diff
@@ -0,0 +1,279 @@
+
+	use Order_Info;
+
+	const DEFAULTS = [
```
There are many cases where this should be `DEFAULT_PARAMS`, but is just `DEFAULTS` instead. Let's pick one, and stick with it. I'd prefer `DEFAULT_PARAMS` as it's already this way within the base requests PR, but I don't have a strong opinion against `DEFAULTS`.

RadoslavGeorgiev commented on includes/core/server/request/class-paginated.php:
```diff
@@ -0,0 +1,140 @@
+	 */
+	public function is_filter_key_mutable( string $key ) {
+		// In most cases, the default parameters are keys that cannot be changed via filters.
+		return ! array_key_exists( $key, self::DEFAULTS );
```
Is there a strong reason why defaults are something that should be immutable here? If so, I'd just declare the same keys as in the `DEFAULT_PARAMS` array info `IMMUTABLE_PARAMS`, instead of introducing a new method.

zmaglica replied:
Make sense. I used `IMMUTABLE_PARAMS` constant to achieve the same results.

RadoslavGeorgiev commented on includes/core/server/request/class-list-transactions.php:
```diff
@@ -0,0 +1,279 @@
+	const DEFAULTS = [
+		'sort'       => 'date',
+		'direction'  => 'desc',
+		'deposit_id' => null,
```
By introducing an empty deposit ID here, we're sending a blank parameter to the server with every request. If it's not required, let's remove the default value here.

zmaglica replied:
Default Deposit have been removed.

RadoslavGeorgiev commented on includes/core/server/request/class-list-transactions.php:
```diff
@@ -0,0 +1,279 @@
+	 * @return void
+	 * @throws \WCPay\Core\Exceptions\Server\Request\Invalid_Request_Parameter_Exception
+	 */
+	public function set_date_after( string $date_after ) {
```
`set_date_after` and `set_date_between` appear a few times throughout the code. Could we maybe consolidate them into a trait?

<img width="468" alt="image" src="https://user-images.githubusercontent.com/5311119/208420973-cc8dc04c-25b1-44b6-b1b8-b64ce986d78c.png">

zmaglica replied:
A new trait has been added.

RadoslavGeorgiev commented on includes/core/server/request/class-list-transactions.php:
```diff
@@ -0,0 +1,279 @@
+	 */
+	public function set_date_between( array $date_between ) {
+		// Add only if it's not empty.
+		if ( $date_between ) {
```
Since we're validating dates in the other methods, let's do it here as well. Considering the parameter is an array with numeric indexes (0 and 1), how about something like this?

```suggestion
		if ( is_array( $date_between ) && ! empty( $date_between ) ) {
			foreach ( $date_between as $date ) {
				$this->validate_date( $date );
			}
			$this->set_param( 'date_between', $date_between );
		}
```

zmaglica replied:
This is refactored and moved into date params trait.

RadoslavGeorgiev commented on includes/core/server/request/class-list-transactions.php:
```diff
@@ -0,0 +1,279 @@
+	 * @return void
+	 */
+	public function set_search( string $search ) {
+		$this->set_param( 'search', $search );
```
This one is incorrect. If you take a look on the server, you'll see that `search` should be an array: https://github.com/Automattic/woocommerce-payments-server/blob/ce757b31f600ba0d2ee582978682856bf24031ce/server/wp-content/rest-api-plugins/endpoints/wcpay/class-transactions-controller.php#L88-L92

Also, there is a missing `WC_Payments_Utils::map_search_orders_to_charge_ids` call, which is otherwise present in the API client method: https://github.com/Automattic/woocommerce-payments/blob/7692800948ef80d3118c5593c32ee6776ea51022/includes/wc-payment-api/class-wc-payments-api-client.php#L666-L668

zmaglica replied:
This is fixed.

RadoslavGeorgiev commented on includes/core/server/request/class-list-transactions.php:
```diff
@@ -0,0 +1,279 @@
+		if ( ! is_null( $date_between_filter ) ) {
+			$date_between_filter = array_map(
+				function ( $transaction_date ) use ( $user_timezone, $wcpay_request ) {
+					return $wcpay_request->format_transaction_date_with_timestamp( $transaction_date, $user_timezone );
```
Even if we set the return type of `parent::from_rest_request()` to be `static`, it's still being returned as `Paginated`, which does not have `format_transaction_date_with_timestamp`, causing linting errors.

What do you think about making the method static as well? It does not need any object-specific functionality.

zmaglica replied:
Good point. Made this function static.

RadoslavGeorgiev reviewed (changes requested):
I did a quick pass, focusing my testing on the list transactions endpoint. I'll review tests later, but so far it's looking pretty good :)

zmaglica committed:
Code review fixes. Fixed naming issue with defaults and removed unusued code

zmaglica committed:
Added date param trait

zmaglica committed:
Added date between validation

zmaglica committed:
Converted paginated class to abstract class

zmaglica committed:
Used immutable params for filters validation override check

zmaglica committed:
Fixed search in list transactions

zmaglica committed:
Changed format transaction function into static one

zmaglica committed:
Fixed small nitpicks regarding variable names and responses

RadoslavGeorgiev committed:
Fixing a Psalm warning

RadoslavGeorgiev commented on includes/core/server/request/class-paginated.php:
```diff
@@ -0,0 +1,131 @@
+
+		foreach ( $filters as $key => $value ) {
+			// Nullable filters are not needed and mutable keys can only be changed with this method.
+			if ( null !== $value && ! in_array( $key, self::IMMUTABLE_PARAMS, true ) ) {
```
Thanks for switching from default parameters to an immutable declaration. With that in mind though, I don't see any reason to keep the `is_array( ... )` check here. If it's needed for whatever reason, I don't see overwriting paging parameters as mission-critical, so let consumers overwrite them :)

RadoslavGeorgiev reviewed (changes requested):
So close! One small nitpick, and I believe we'll be done with this one :)

zmaglica committed:
Removed in array check for filters

zmaglica committed:
Removed ununsed imports and lint fixes

zmaglica committed:
Merge branch 'dev/create-and-setup-intent-class' of github.com:Automattic/woocommerce-payments into dev/create-and-setup-intent-class

zmaglica committed:
Removed unnecessary empty lines

zmaglica committed:
Removed unused elements from update intention class

zmaglica committed:
Removed unnecessary metadata assrts from UPE payment gateway test, gateway test and update intention request

zmaglica committed:
Removed set selected upe payment method types from update intent to gateway class

zmaglica committed:
Removed asserts that are not needed

zmaglica committed:
Removed ununsed imports and lint fixes

RadoslavGeorgiev reviewed (commented):
All changes in the PR look good, let's make sure all tests are passing once updated with the contents of the `dev/create-and-setup-intent-class` :)

zmaglica committed:
Fixed tests

RadoslavGeorgiev committed:
Merge branch 'dev/create-and-setup-intent-class' into dev/paginated-request-class

RadoslavGeorgiev committed:
Removing filter immutability tests.

At this point filters for paginated requests are on the same level as other parameters, hence we should no longer check if they are immutable. This commit removes elements from the `$filters` array, and the associated assertion.

zmaglica committed:
Fixed issue with tests on PHP 7.2

zmaglica committed:
Fixed issue with deposit id

zmaglica committed:
Changed return dockblock to static from Paginated

RadoslavGeorgiev committed:
Fix psalm warnings

RadoslavGeorgiev committed:
Merge branch 'dev/paginated-request-class' of github.com:Automattic/woocommerce-payments into dev/paginated-request-class

zmaglica committed:
Merged core branch
